File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String str;

        while ((str = br.readLine()) != null && !str.equals(".")) {
        	Stack<Character> stack = new Stack<>();

            for(int i = 0; i < str.length(); i++) {
                char c = str.charAt(i);
                if(c == '(' || c == ')' || c == '[' || c == ']') {
                    if(!stack.empty() && stack.peek() == '(' && c == ')')
                        stack.pop();
                    else if (!stack.empty() && stack.peek() == '[' && c == ']')
                    	stack.pop();
                    else
                        stack.push(c);
                }
             }
            
            if(stack.empty())
              System.out.println("yes");
            else
              System.out.println("no");
        }
    }
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
input = sys.stdin.readline

while True:
    command = input().rstrip()
    stack = []

    if command == '.':
        break
    
    for i in command:
        if i == '(' or i == '[':
            stack.append(i)
        elif i == ')':
            if len(stack) != 0 and stack[-1] == '(':
                stack.pop()
            else:
                stack.append(i)
        elif i == ']':
            if len(stack) != 0 and stack[-1] == '[':
                stack.pop()
            else:
                stack.append(i)
                
    if len(stack) == 0:
        print('yes')
    else:
        print('no')